tokio/src/runtime/scheduler/multi_thread/worker.rs

```diff
@@ -494,11 +497,21 @@ impl Context {
                     None => return Ok(core),
                 };
 
-                // Polling a task doesn't necessarily consume any budget, if it
-                // doesn't use any Tokio leaf futures. To prevent such tasks
-                // from using the lifo slot in an infinite loop, we consume an
-                // extra unit of budget between each iteration of the loop.
-                coop::consume_one();
+                // In ping-ping style workloads where task A notifies task B,
+                // which notifies task A again, continuously prioritizing the
+                // LIFO slot can cause starvation as these two tasks will
+                // repeatedly schedule the other. To mitigate this, we limit the
+                // number of times the LIFO slot is prioritized.
+                if lifo_polls >= MAX_LIFO_POLLS_PER_TICK {
+                    core.run_queue.push_back_or_overflow(
+                        task,
+                        self.worker.inject(),
+                        &mut core.metrics,
+                    );
+                    return Ok(core);
+                }
+
+                lifo_polls += 1;
 
                 if coop::has_budget_remaining() {
```

Comment:
nit, TIOLI: it seems like we could combine  some of these branches, like
```suggestion
                // In ping-ping style workloads where task A notifies task B,
                // which notifies task A again, continuously prioritizing the
                // LIFO slot can cause starvation as these two tasks will
                // repeatedly schedule the other. To mitigate this, we limit the
                // number of times the LIFO slot is prioritized.
                let lifo_polls_remaining = lifo_polls >= MAX_LIFO_POLLS_PER_TICK;

                if lifo_polls_remaining && coop::has_budget_remaining() {
                    lifo_polls += 1;
```